### PR TITLE
Add Ledger Objects and `Metadata`

### PR DESCRIPTION
**Pull request by LimpidCrypto**
## High Level Overview of Change

This PR adds [`Ledger Objects`](https://xrpl.org/ledger-object-types.html#ledger-object-types) to have a more pythonic, object orientated way of structuring data (the objects themselves, transactions, ...). It will also add a `Metadata` model representing a transaction's metadata. The models can be derived directly from a dictionary.
There is already a `TransactionMetadata` model (which is a bit messy), that basically also represents a transaction's metadata. The issue with that model is that it has no real checks if the fields got derived correctly, if there were required fields missing, etc. It also only predefines ledger object fields currently needed for this repository, which by far aren't all possible fields.
Having actual models for Ledger Objects and Metadata also delivers better typing of the code.

There are separate models for Ledger Objects when represented in a transaction's metadata (`MD...Fields`). The reason behind this is that nodes can have `FinalFields` or `PreviousFields` where some of the normally `REQUIRED` fields are not included. Like if an `Offer` object gets changed in the `PreviousFields` are only `TakerGets` and `TakerPays` defined, all other fields are missing there, because they are defined in `FinalFields` or a level higher. That's also why in those models all fields are `Optional`

The models are intendedly not exported to the `models/__init__.py` because models like `DepositPreauth` would collide with the transaction type.

### Context of Change

#205

### Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Test Plan

- [x] Test `from_dict`
- [x] Test `from_xrpl`
- [x] Test `to_dict`
- [ ] Test errors

## Tasks

- [x] Add [ledger header](https://xrpl.org/ledger-header.html) models
- [ ] Remove old metadata representation as soon as there is confidence that this PR has the chance to get merged. The transaction parsers currently use the old models, those have to be replaced.
- [x] Put the json test transactions into a separate directory

### CODE REVIEW

File: xrpl/models/ledger_objects/account_root.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    wallet_locator: Optional[str] = None
+    wallet_size: Optional[int] = None
+    ledger_entry_type: LedgerEntryType = field(
+        default=LedgerEntryType.ACCOUNT_ROOT, init=False
+    )
+
+
+@require_kwargs_on_init
+@dataclass(frozen=True)
+class MDAccountRootFields(LedgerObject):
```
Comment (mvadari) [inline xrpl/models/ledger_objects/account_root.py:47]: IMO the metadata version of these objects doesn't need to be included - we don't include it in the TypeScript models in xrpl.js, the metadata fields are just strings.

Comment (LimpidCrypto) [reply xrpl/models/ledger_objects/account_root.py:47]: So you would prefer just to have let's say a ˋAccountRootˋ model with all fields to be optional to also use them in the ˋMetadataˋ?

Comment (mvadari) [reply xrpl/models/ledger_objects/account_root.py:47]: The opposite, actually - see: https://github.com/XRPLF/xrpl.js/blob/main/packages/xrpl/src/models/transactions/metadata.ts

Comment (LimpidCrypto) [reply xrpl/models/ledger_objects/account_root.py]: Seems responsible :)

**Review by mvadari — COMMENTED** [review]
The metadata and ledger objects should be in separate PRs.

**Comment by JST5000** [comment]
Sorry, somehow I completely missed this PR, I'll aim to review it by Monday :)

**Comment by LimpidCrypto** [comment]
> Sorry, somehow I completely missed this PR, I'll aim to review it by Monday :)

All good, I need to find time to continue this PR first anyway :D

**Review by mvadari — COMMENTED** [review]
Perhaps this PR should be split into two: one for the ledger objects and one for the metadata. I suspect the metadata part will involve a lot more review.

**Comment by LimpidCrypto** [comment]
> Perhaps this PR should be split into two: one for the ledger objects and one for the metadata. I suspect the metadata part will involve a lot more review.

I agree 👍🏼 I will close this PR for now and hope I get to open a PR for the ledger objects this weekend :)

**Pull request closed without merging**